### ISSUE TITLE
-g3 .wasm size after strip-debug does not equal size with -g0

### ISSUE THREAD

**Issue by bkotsopoulossc** [emscripten-core/emscripten#15269]
Related to some discussion in https://github.com/emscripten-core/emscripten/issues/13084

Let's say I build a wasm binary with `-g0`, and measure the size. Then I build the binary with the same flags, but with `-g3`, and then take that binary, and strip out debug information, I expect the resulting size to be the same as the first binary I built. But it doesn't seem to be. I am not sure if this is because the debug mode turns off certain compiler optimizations, or if it means that extra data gets generated in the binary in sections other than the debug one.

I am using the [bazel SDK](https://github.com/emscripten-core/emsdk/tree/main/bazel) for this, but it is unclear if that is part of the problem or not. There is already a small test app [here](https://github.com/emscripten-core/emsdk/tree/main/bazel/test_external) that can reproduce the problem, although the size difference is much bigger when using real projects with more code.

Steps:

Build and link with -Oz
```
cd bazel/test_external
% bzl build --copt=-Oz --linkopt=-Oz //:hello-world-wasm
% ls -al bazel-bin/hello-world-wasm/hello-world.wasm    
120959  8 Oct 16:11 bazel-bin/hello-world-wasm/hello-world.wasm
```

Build and link with -Oz -g3
```
% bzl build --copt=-g3 --linkopt=-g3 --copt=-Oz --linkopt=-Oz //:hello-world-wasm
% ls -al bazel-bin/hello-world-wasm/hello-world.wasm
2217363  8 Oct 16:13 bazel-bin/hello-world-wasm/hello-world.wasm
```

Strip using `objcopy`, in the same way as used [here](https://github.com/emscripten-core/emscripten/blob/main/tools/building.py#L1140), as called from [here](https://github.com/emscripten-core/emscripten/blob/main/tools/building.py#L1164)

```
% cp bazel-bin/hello-world-wasm/hello-world.wasm debug.wasm
% bzl run @emscripten_bin_mac//:bin/llvm-objcopy -- --remove-section='.debug*' debug.wasm 
% ls -al debug.wasm
% ls -al debug.wasm 
309399  8 Oct 16:24 debug.wasm
```

We went from 120KB at `-g0` to 309KB after stripping `-g3` on this empty test app.

Is this a bug or expected behaviour? How come this happens? Are there additional sections that can be stripped after that fact to get back down to the original size?

CC @sbc100

**Comment by bkotsopoulossc** [comment]
@sbc100 we get closer, but seems like not all the way:

(After the -g3 -O3 build)
```
% cp bazel-bin/hello-world-wasm/hello-world.wasm debug.wasm
% bazel run @emscripten_bin_mac//:bin/llvm-objcopy -- --strip-all debug.wasm
% ls -al debug.wasm 
142352 12 Oct 19:41 debug.wasm
```

So 120KB to 142KB

**Comment by bkotsopoulossc** [comment]
Had a chance to try this on emcc directly, using the same test file from the bazel project.

128KB with -Oz
```
% emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.0.0 (3fd52e107187b8a169bb04a02b9f982c8a075205)
% emcc -Oz bazel/hello-world/hello-world.cc -o hello-world.wasm 
% ls -al hello-world.wasm 
128952 30 Dec 17:32 hello-world.wasm
```

149KB stripping a -Oz -g
```
% emcc -g3 -Oz bazel/hello-world/hello-world.cc -o hello-world.wasm
% ./upstream/bin/llvm-objcopy --strip-all hello-world.wasm stripped.wasm
% ls -al stripped.wasm  
149156 30 Dec 17:34 stripped.wasm
```

**Comment by bkotsopoulossc** [comment]
If we use objdump, there seems to be meaningful differences in the code sections:
```
 % wasm-objdump -h oz.wasm

oz.wasm:	file format wasm 0x1

Sections:

     Type start=0x0000000b end=0x00000196 (size=0x0000018b) count: 48
   Import start=0x00000199 end=0x000002dd (size=0x00000144) count: 9
 Function start=0x000002e0 end=0x00000532 (size=0x00000252) count: 592
    Table start=0x00000534 end=0x0000053b (size=0x00000007) count: 1
   Memory start=0x0000053d end=0x00000543 (size=0x00000006) count: 1
   Global start=0x00000545 end=0x0000054e (size=0x00000009) count: 1
   Export start=0x00000550 end=0x000005bd (size=0x0000006d) count: 7
     Elem start=0x000005c0 end=0x00000837 (size=0x00000277) count: 1
     Code start=0x0000083b end=0x0001cc9f (size=0x0001c464) count: 592
     Data start=0x0001cca2 end=0x0001f7b8 (size=0x00002b16) count: 52
```

```
% wasm-objdump -h oz_g_stripped.wasm

oz_g_stripped.wasm:	file format wasm 0x1

Sections:

     Type start=0x0000000e end=0x000001d2 (size=0x000001c4) count: 58
   Import start=0x000001d8 end=0x0000031c (size=0x00000144) count: 9
 Function start=0x00000322 end=0x00000917 (size=0x000005f5) count: 1523
    Table start=0x0000091d end=0x00000924 (size=0x00000007) count: 1
   Memory start=0x0000092a end=0x00000930 (size=0x00000006) count: 1
   Global start=0x00000936 end=0x0000093f (size=0x00000009) count: 1
   Export start=0x00000945 end=0x000009b2 (size=0x0000006d) count: 7
     Elem start=0x000009b8 end=0x00000c6e (size=0x000002b6) count: 1
     Code start=0x00000c74 end=0x00021b88 (size=0x00020f14) count: 1523
     Data start=0x00021b8e end=0x000246a4 (size=0x00002b16) count: 52
```

**Comment by bkotsopoulossc** [comment]
Did this same comparison for -O3 and -O0, and the difference still exists, but its more pronounced with different optimization levels.
```
161419 30 Dec 17:46 o0.wasm
2290547 30 Dec 17:46 o0_g.wasm
164065 30 Dec 17:45 o0_g_stripped.wasm
144574 30 Dec 17:44 o3.wasm
2260559 30 Dec 17:44 o3_g.wasm
150190 30 Dec 17:44 o3_g_stripped.wasm
128952 30 Dec 17:36 oz.wasm
2258757 30 Dec 17:36 oz_g.wasm
149156 30 Dec 17:36 oz_g_stripped.wasm
```

**Comment by bkotsopoulossc** [comment]
If I use the -O0 example, and compare the output from `-v`, I get the following:
- `emcc` commands are identical (except the object file path)
- `clang-14` command has extra flags for the debug build: `-debug-info-kind=constructor -dwarf-version=4`
- `wasm-ld` has flags `--strip-debug` for the non-debug build
- `wasm-emscripten-finalize` has flags `-g --dwarf` for the debug build
- `llvm-objcopy` has flags `--remove-section=.debug*` for the non-debug build

So there aren't additional libs being linked in or anything

**Comment by bkotsopoulossc** [comment]
@sbc100 any pointers on things to look into next here?

**Comment by sbc100** [comment]
I did some digging and it looks like that difference is due to the `-g` flag being passed to `wasm-opt` preventing `wasm-opt` from removing certain functions.  

The different wasm-opt commands (from -g build and no--g builds) are :

```
wasm-opt --strip-dwarf --post-emscripten -Oz --low-memory-unused --zero-filled-memory --strip-debug --strip-producers withg.wasm -o withg.wasm --mvp-features
wasm-opt --post-emscripten -Oz --low-memory-unused --zero-filled-memory --strip-producers withg.wasm -o withg.wasm -g --mvp-features
```

It seems that including the `-g` here prevent wasm-opt from reducing the number of functions from 32 down to 13 (for hello_world.c).

@kripken do you know what is going on here?

**Comment by sbc100** [comment]
My best guess here is that passing `-g` to wasm-opt, in the presence of debug info is causing wasm-opt to skip certain passes (such as inlining), because it can't mutate the dwarf info in the way that would be needed to support this.    This means that adding `-g` to an `-Oz` does indeed effect the output binary (in in fairly major way).   This is certainly not good and I'm not sure what we can do about it.

Possible solution:

1. Error or warn when folks use `-Oz/etc` with `-g` because these flags are (from a wasm-opt POV) mutually exclusive.
2. Work on improving our dwarf re-writing in binaryen.

**Comment by kripken** [comment]
Looks like these passes are marked as invalidating DWARF in binaryen, and are not run when that is enabled:
```
CoalesceLocals
DeadArgumentElimination
DuplicateFunctionElimination
Inlining
LocalCSE
MergeLocals
SSAify
```
That is a current limitation atm. If we change how we handle DWARF, that might be improved. But we don't have a plan for that yet. Improving the DWARF rewriting would be a long-term project. Other options are to remove that code and replace it with another method of updating debug info.

https://github.com/WebAssembly/binaryen/pull/4457 may also have some small effect here, but I think it'll be 99% due to those pass disablings.

It's tricky to error on `-Oz -g` because that would be a surprising breaking change. Warning perhaps, for now?

**Comment by bkotsopoulossc** [comment]
Thanks for the info, very valuable. Just FYI, from my tests, I observed the same thing for -Os, -O3, and -O2.

**Comment by bkotsopoulossc** [comment]
As far as I know, native build toolchains don't suffer from this issue. I'm not sure how they work around this and are able to optimize and emit debug symbols without affecting code size.

This does feel like a fairly big issue, to not be able to symbolicate a crash/error trace from a production binary, without taking a pretty big size overhead. 

Are there any other options as workarounds? How bad would the dwarf info be corrupted if those passes were enabled?

**Comment by sbc100** [comment]
Native toolchains don't generally do post-link optimization like wasm-opt does, which is why they don't have this issue.

One solution if you just want to be able to symbolicate a crash report is to use `--profiling-funcs` or `--emit-source-map` instead of `-g`.      In both cases you will end up with enough information to do function-level symbolicate'ing and neither of these flags should effect wasm-opt (since its only DWARF information that causes wasm-opt to back off).    `--profiling-funcs` will emit a name section in your wasm binary, and `--emit-source-map` will write a separate file that maps function indexes to names.

**Comment by sbc100** [comment]
I agree this is serious issue, and I think we will at very leave have to add a warning message when folks do `-O` + `-g`.

@dschuff WDYT?  Is a warning a reasonable interim fix?

**Comment by bkotsopoulossc** [comment]
> One solution if you just want to be able to symbolicate a crash report is to use `--profiling-funcs` or `--emit-source-map` instead of `-g`. In both cases you will end up with enough information to do function-level symbolicate'ing and neither of these flags should effect wasm-opt (since its only DWARF information that causes wasm-opt to back off). `--profiling-funcs` will emit a name section in your wasm binary, and `--emit-source-map` will write a separate file that maps function indexes to names.

Thanks for the suggestion, I am trying to understand it better. Do both of these options involve actually shipping a name section in the prod binary? The name section is huge, so I don't think that will be feasible. Is the idea to strip the name section but save it and use it offline for symbolication? I'm having a hard time understanding how that would work

**Comment by sbc100** [comment]
`--emit-symbol-map` writes the function name to an external file so is probably the best solution for your use case.

**Comment by bkotsopoulossc** [comment]
Got it, thanks. That map appears to have the function name symbol, but do we have another mechanism to associate the function with the source file?

And I guess a limitation of this is you're getting function names but not line numbers right?

**Comment by bkotsopoulossc** [comment]
We raced again :) 

So if we want line level granularity without shipping a name section, maybe source maps are the way to go then?

And yeah, I think that we get 95% of the value by having accurate line tables in debug symbols for optimized binaries. The use cases for the rest of the symbols aren't nearly as common

**Comment by sbc100** [comment]
I wonder if binaryen can allow all optimization if we include only limited DWARF info such as  `-gline-tables-only` ?   @kripken WDYT, would that be a much simpler re-writing problem?

**Comment by dschuff** [comment]
Yeah, that was basically what I meant. Binaryen keeps track of the line location for each instruction, and can probably do a reasonable job of copying it (e.g. for new instructions that it creates), even when doing things that modify or destroy variable values (like optimizations on locals).

**Comment by sbc100** [comment]
Sorry I missed that part of your previous comment.  Great minds..

**Comment by dschuff** [comment]
I've uploaded #16110 to fix #13084, from whence this bug was forked.
While looking at that, I also realized that the current behavior leaves the "name" section in the original binary along with `external_debug_info`. I wonder if this is what we want? On the one hand one value of the name section is that it's recognized by the engine, and for that use case it needs to be in the binary that runs. But on the other hand if you actually want a small production binary you'll need to strip the main wasm again.

A related thing I also just noticed that @bkotsopoulossc said in that thread:
> The usefulness of separate-dwarf is that you can build your -g0 and -g3 artifacts in one swoop

That's actually not the reason that `-gseparate-dwarf` exists. The real reason is when you want to do live debugging on a binary with 40 MB of code and 1GB of debug info (a realistic number!) the VM has trouble loading the binary despite the fact that most of it is unneeded for actual running. So `-gseparate-dwarf` fixes that problem (an even better solution is split-dwarf but the debugger wasn't ready for that when we implemented separate-dwarf). But the point is that it's here to support the live debugging use case, rather than ship-a-minimal-binary-symbolize-crashes-from-the-field use case (another thing of note is that those use cases usually use different optimization flags). So in principle maybe it's not a huge problem that the way to do the latter is the same as on other platforms (i.e. build with full-optimization flags, then copy and strip).

Having said all that, what we said above still holds; it's an unfortunate problem that the best optimization breaks debug info. Currently it means that you can't get correct symbolization of the best-optimized bits with DWARF (although I think you can with a source map). And even once we have the "quick" fix of correct DWARF line tables for full optimization, it still means that you can't really debug your fully-optimized exact production bits with a debugger.

**Comment by bkotsopoulossc** [comment]
> I wonder if this is what we want? On the one hand one value of the name section is that it's recognized by the engine, and for that use case it needs to be in the binary that runs. But on the other hand if you actually want a small production binary you'll need to strip the main wasm again.

My hunch is that its more common when using separate-dwarf to want to get rid of name, but it's not too bad to have to do it as a post-process step.

> The real reason is when you want to do live debugging on a binary with 40 MB of code and 1GB of debug info (a realistic number!) the VM has trouble loading the binary despite the fact that most of it is unneeded for actual running

In my experience with recent chrome versions, as long as you use streaming compilation, big binaries, even >1GB load just fine if symbols and code are in the same binary. From the console logs (the debug extension spits out blue logs when it starts and finishes loading debug symbols), it seems like the debug symbols get loaded asynchronously and don't block loading of the code. I haven't seen much benefit splitting into the two files when debugging. 

> maybe it's not a huge problem that the way to do the latter is the same as on other platforms (i.e. build with full-optimization flags, then copy and strip)

I agree with that.

> it still means that you can't really debug your fully-optimized exact production bits with a debugger.

Even on other platforms like debugging native binaries, this is a huge problem. With all the inlining that goes on, debugging a heavily optimized binary is very difficult. Very few people are able to do it well.

**Comment by sbc100** [comment]
I added a new warning so folks are aware of this issue if they try to request dwarf info in optimized builds: #15269

**Comment by tommitytom** [comment]
I notice that I still get strack traces in the browser console even when using `-gseparate-dwarf`, and deleting the resulting debug file.  To what extent could someone reverse engineer my .wasm from the additional information remaining after the split?

**Comment by stheophil** [comment]
> I've uploaded #16110 to fix #13084, from whence this bug was forked. While looking at that, I also realized that the current behavior leaves the "name" section in the original binary along with `external_debug_info`. I wonder if this is what we want? On the one hand one value of the name section is that it's recognized by the engine, and for that use case it needs to be in the binary that runs. But on the other hand if you actually want a small production binary you'll need to strip the main wasm again.

I have just encountered this issue and spent 2 hours trying to find out if there is a bug in my build configuration. I am strongly in favor of stripping the names. With -g3 and C++ that makes heavy use of templates, the names section can be absolutely huge. 

When compiling with -g2, bloaty reports

>      FILE SIZE        VM SIZE    
>  --------------  -------------- 
>   92.1%  4.26Mi   NAN%       0    name
>    6.8%   321Ki   NAN%       0    Code
>    1.0%  46.3Ki   NAN%       0    Data
>    0.0%  1.81Ki   NAN%       0    Function
>    0.0%  1.23Ki   NAN%       0    Element
>    0.0%     510   NAN%       0    Type
>    0.0%     256   NAN%       0    Import
>    0.0%      72   NAN%       0    Export
>    0.0%      11   NAN%       0    Global
>    0.0%       9   NAN%       0    Table
>    0.0%       8   NAN%       0    Memory
>    0.0%       8   NAN%       0    [WASM Header]
>  100.0%  4.62Mi 100.0%       0    TOTAL

Bad, but shipping a 5Mb binary is at least feasible. When compiling with -gseparate-dwarf:

>     FILE SIZE        VM SIZE    
>  --------------  -------------- 
>   99.5%   156Mi   NAN%       0    name
>    0.4%   678Ki   NAN%       0    Code
>    0.0%  46.3Ki   NAN%       0    Data
>    0.0%  24.6Ki   NAN%       0    Function
>    0.0%  1.79Ki   NAN%       0    Element
>    0.0%  1.13Ki   NAN%       0    Import
>    0.0%     805   NAN%       0    Type
>    0.0%     314   NAN%       0    Export
>    0.0%      52   NAN%       0    external_debug_info
>    0.0%      15   NAN%       0    Global
>    0.0%      13   NAN%       0    Table
>    0.0%      12   NAN%       0    Memory
>    0.0%       8   NAN%       0    [WASM Header]
>  100.0%   157Mi 100.0%       0    TOTAL

This is ridiculous of course. The binary is the same size as the dwarf module file.

**Comment by stheophil** [comment]
Also shows issue https://github.com/emscripten-core/emscripten/issues/15269

The code section is twice the size when compiling with -gseparate-dwarf

**Comment by sbc100** [comment]
You can run `emstrip` or `llvm-strip` to remove the resulting name section.

It sounds like we should not be including it at all in the `-gseparate-dwarf` case.  If anyone wants to explicitly preserve the name section we already have the `--profiling-funcs` flag which forcing the name section to be included.
